src/webgpu.c

```diff
@@ -0,0 +1,198 @@
+        fromWaylandSurface.chain.sType = WGPUSType_SurfaceDescriptorFromWaylandSurface;
+#  endif
+        fromWaylandSurface.chain.next = NULL;
+        fromWaylandSurface.display = wayland_display;
+        fromWaylandSurface.surface = wayland_surface;
+
+        WGPUSurfaceDescriptor surfaceDescriptor;
+        surfaceDescriptor.nextInChain = &fromWaylandSurface.chain;
+        surfaceDescriptor.label = NULL;
+
+        return wgpuInstanceCreateSurface(instance, &surfaceDescriptor);
+    }
+#endif // GLFW_EXPOSE_NATIVE_WAYLAND
+
+#ifdef GLFW_EXPOSE_NATIVE_COCOA
+    case GLFW_PLATFORM_COCOA: {
+        id metal_layer = [CAMetalLayer layer];
+        NSWindow* ns_window = glfwGetCocoaWindow(window);
+        [ns_window.contentView setWantsLayer : YES] ;
+        [ns_window.contentView setLayer : metal_layer] ;
+
+#  ifdef WEBGPU_BACKEND_DAWN
+        WGPUSurfaceSourceMetalLayer fromMetalLayer;
+        fromMetalLayer.chain.sType = WGPUSType_SurfaceSourceMetalLayer;
+#  else
+        WGPUSurfaceDescriptorFromMetalLayer fromMetalLayer;
+        fromMetalLayer.chain.sType = WGPUSType_SurfaceDescriptorFromMetalLayer;
+#  endif
+        fromMetalLayer.chain.next = NULL;
+        fromMetalLayer.layer = metal_layer;
+
+        WGPUSurfaceDescriptor surfaceDescriptor;
+        surfaceDescriptor.nextInChain = &fromMetalLayer.chain;
+        surfaceDescriptor.label = NULL;
+
+        return wgpuInstanceCreateSurface(instance, &surfaceDescriptor);
+    }
+#endif // GLFW_EXPOSE_NATIVE_COCOA
+
+#ifdef GLFW_EXPOSE_NATIVE_WIN32
+    case GLFW_PLATFORM_WIN32: {
+        HWND hwnd = glfwGetWin32Window(window);
+        HINSTANCE hinstance = GetModuleHandle(NULL);
+
+#  ifdef WEBGPU_BACKEND_DAWN
+        WGPUSurfaceSourceWindowsHWND fromWindowsHWND;
+        fromWindowsHWND.chain.sType = WGPUSType_SurfaceSourceWindowsHWND;
+#  else
+        WGPUSurfaceDescriptorFromWindowsHWND fromWindowsHWND;
+        fromWindowsHWND.chain.sType = WGPUSType_SurfaceDescriptorFromWindowsHWND;
+#  endif
+        fromWindowsHWND.chain.next = NULL;
+        fromWindowsHWND.hinstance = hinstance;
+        fromWindowsHWND.hwnd = hwnd;
+
+        WGPUSurfaceDescriptor surfaceDescriptor;
+        surfaceDescriptor.nextInChain = &fromWindowsHWND.chain;
+        surfaceDescriptor.label = NULL;
+
+        return wgpuInstanceCreateSurface(instance, &surfaceDescriptor);
+    }
+#endif // GLFW_EXPOSE_NATIVE_X11
+
+#ifdef GLFW_EXPOSE_NATIVE_EMSCRIPTEN
+    case GLFW_PLATFORM_EMSCRIPTEN: {
+#  ifdef WEBGPU_BACKEND_DAWN
+        WGPUSurfaceSourceCanvasHTMLSelector_Emscripten fromCanvasHTMLSelector;
+        fromCanvasHTMLSelector.chain.sType = WGPUSType_SurfaceSourceCanvasHTMLSelector_Emscripten;
+#  else
+        WGPUSurfaceDescriptorFromCanvasHTMLSelector fromCanvasHTMLSelector;
+        fromCanvasHTMLSelector.chain.sType = WGPUSType_SurfaceDescriptorFromCanvasHTMLSelector;
+#  endif
+        fromCanvasHTMLSelector.chain.next = NULL;
+        fromCanvasHTMLSelector.selector = "canvas";
```

Comment:
I think the best would be to have something like an API to create a GLFW window from an existing window handle, then pass a handle to a specific canvas as the existing window

Reply:
I agree it is a bad idea to have hardcoded the canvas selector, but I am not sure what is the most idiomatic way to have this emscripten-only option exposed to the API, nor how GLFW chooses its target canvas in general (is there a way to retrieve a target canvas from the window handle?

Reply:
@eliemichel The thing is, this code is kind of orthogonal to anything GLFW does, really.  I know [in your WebGPU tutorial, you create a GLFW window](https://eliemichel.github.io/LearnWebGPU/getting-started/opening-a-window.html), but it's not actually needed at all in the context of a canvas selector.  When doing this in emscripten for the web, you can just omit the GLFW window creation code entirely.  It's enough to do the following on its own, to start rendering:

```cpp
      wgpu::SurfaceDescriptorFromCanvasHTMLSelector surface_descriptor_from_canvas;
      surface_descriptor_from_canvas.selector = "#canvas";

      wgpu::SurfaceDescriptor surface_descriptor{
        .nextInChain{&surface_descriptor_from_canvas},
        .label{"Canvas surface"},
      };
      auto webgpu.surface{instance.CreateSurface(&surface_descriptor)};
```
(example using emscripten's webgpu_cpp.h)

Obviously GLFW might still then be useful for processing input etc, but unless you're going to call any GLFW functions that take `window` as a parameter, there may be no need to ever call `glfwCreateWindow`.  So in a way, it doesn't really matter how GLFW handles canvases in OpenGL mode, because none of that applies to us with WebGPU.  After all, we're not getting a GL context for that GLFW window, or interacting with it in any way as far as WebGPU rendering is concerned.

I think having this function take an argument for selector name would be the best solution to make this flexible enough to be useful.  I think any tighter coupling with other GLFW features would just be unnecessarily limiting at this stage.